core/src/precompile.rs

```diff
@@ -0,0 +1,270 @@
+use alloc::{boxed::Box, collections::BTreeMap, sync::Arc, vec::Vec};
+use core::error::Error;
+
+use miden_crypto::{Felt, Word, hash::rpo::Rpo256};
+use winter_math::FieldElement;
+
+use crate::{
+    EventId,
+    utils::{ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable},
+};
+
+// PRECOMPILE REQUEST
+// ================================================================================================
+
+/// Represents a single precompile request consisting of an event ID and byte data.
+///
+/// This structure encapsulates the call data for a precompile operation, storing
+/// the raw bytes that will be processed by the precompile function.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PrecompileRequest {
+    /// Event ID identifying the type of precompile operation
+    pub event_id: EventId,
+    /// Raw byte data for the precompile operation
+    pub data: Vec<u8>,
+}
+
+impl Serializable for PrecompileRequest {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        self.event_id.as_felt().write_into(target);
+        self.data.write_into(target);
+    }
+}
+
+impl Deserializable for PrecompileRequest {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let event_id = EventId::from_felt(Felt::read_from(source)?);
+        let data = Vec::<u8>::read_from(source)?;
+        Ok(Self { event_id, data })
+    }
+}
+
+// PRECOMPILE COMMITMENT
+// ================================================================================================
+
+/// Represents the result of a verified precompile computation.
+///
+/// This structure contains both the tag (which includes metadata like event ID)
+/// and the commitment word that represents the verified computation result.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PrecompileCommitment {
+    /// Tag containing metadata including the event ID in the first element
+    pub tag: Word,
+    /// Commitment word representing the verified computation result
+    pub commitment: Word,
+}
+
+impl PrecompileCommitment {
+    /// Returns the concatenation of tag and commitment as field elements.
+    pub fn to_elements(&self) -> [Felt; 8] {
+        let words = [self.tag, self.commitment];
+        Word::words_as_elements(&words).try_into().unwrap()
+    }
+}
+
+// PRECOMPILE VERIFIERS REGISTRY
+// ================================================================================================
+
+/// Registry of precompile verifiers.
+///
+/// This struct maintains a map of event IDs to their corresponding verifiers.
+/// It is used to verify precompile requests during proof verification.
+#[derive(Default, Clone)]
+pub struct PrecompileVerifiers {
```

Comment:
Should we call this `PrecompileVerifierRegistry`? Also, I wonder if there is a better term than "verifier" - we usually use "verifier" in the context of verifying ZK proofs - and it may get a bit confusing if we use it here as well.

Reply:
In some sense, these are "verifiers" in terms of the theoretical NP definition, though I would also call it a "decider" since it's performing computation. I'll leave verifier for now until we find a better name.

Reply:
I guess what throws me off a bit is that the verifiers don't actually verify the validity. For example, for Keccak, we are just computing the hash, but the `verify()` function doesn't have enough info to figure out if the computed result is valid (the verification step comes later after we combine all commitments and check the result against the commitment output by the VM).

To me, this feels more like "precompile commitment builder" as it is main job is to convert precompile request into precompile commitment. But that's not a great name either.